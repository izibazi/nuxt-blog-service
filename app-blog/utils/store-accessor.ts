import { Store } from 'vuex'
import { getModule } from 'vuex-module-decorators'

// アプリケーションに必要なStoreを初期化する
function initializeStore(store: Store<any>): void {}

export { initializeStore }
