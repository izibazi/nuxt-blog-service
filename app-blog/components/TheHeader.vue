<template lang="pug">
  div
    el-menu(mode="horizontal" :router="true")
      el-menu-item(index="1" :route="{path: '/'}")
        h1 Nuxt Diary App
      el-menu-item(index="2" :route="{path: '/posts/'}") 投稿
      el-menu-item(index="3" key="user" v-if="user" :route="{path: `/users/${user.id}`}"  style="float: right") {{ user.id }}
      el-menu-item(index="3" key="user" v-else :route="{path: '/'}" style="float: right") ログイン
      el-menu-item(index="4" :route="{path: '/posts/new'}" style="float: right") 新規投稿
</template>

<script lang="ts">
import { Vue, Component } from 'nuxt-property-decorator'
import { MenuItem, Menu } from 'element-ui'
import { authStore } from '~/utils/store-accessor'
import { User } from '~/domain/user'

@Component({
  components: {
    'el-menu': Menu,
    'el-menu-item': MenuItem,
  },
})
export default class TheHeader extends Vue {
  get user(): User | null {
    return authStore.user
  }
}
</script>

<style scoped>
h1 {
  font-size: 1em;
  margin: 0;
}
</style>
