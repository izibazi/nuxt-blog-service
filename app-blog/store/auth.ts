import { Module, VuexModule, Mutation, Action } from 'vuex-module-decorators'
import { IUser } from '~/domain/user'
import { $axios } from '~/utils/api'

@Module({ stateFactory: true, name: 'auth' })
export default class auth extends VuexModule {
  user: IUser | null = null

  get isLoggedIn(): boolean {
    return this.user != null
  }

  @Mutation
  setUser(user: IUser): void {
    this.user = user
  }

  @Action({ rawError: true })
  public async login(id: string) {
    const data = await $axios.$get(`/users/${id}.json`)
    if (!data.id) throw new Error('invalid user')
    this.setUser(data)
  }

  @Action
  public async register(id: string) {
    await $axios.$patch('/users.json', { id })
    await this.login(id)
  }
}
