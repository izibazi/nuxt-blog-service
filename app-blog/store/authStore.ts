import { Module, VuexModule, Mutation, Action } from 'vuex-module-decorators'
import cloneDeep from 'lodash.clonedeep'
import { User } from '~/domain/user'
import { Like } from '~/domain/like'
import { IUserRepository } from '~/domain/iUserRepository'
import { $axios } from '~/utils/api'

@Module({ stateFactory: true, name: 'authStore', namespaced: true })
export default class AuthStore extends VuexModule {
  userRepository!: IUserRepository

  user: User | null = null

  get isLoggedIn(): boolean {
    return this.user != null
  }

  get isLiked() {
    return (postId: string): boolean => {
      return (
        this.user?.likes.find((like) => like.postId === postId) !== undefined
      )
    }
  }

  @Mutation
  setUser(user: User): void {
    this.user = user ? Object.assign({ likes: [] }, user) : null
  }

  @Action({ commit: 'setUser', rawError: true })
  public async login(id: string) {
    if (id == null || id.length === 0)
      throw new Error('Can not login.id is empty!')
    const { data } = await this.userRepository.login(id)
    if (!data?.id) throw new Error('invalid user')
    return data
  }

  @Action({ commit: 'setUser', rawError: true })
  public async register(id: string) {
    id = id.trim()
    if (id.length === 0) throw new Error('Can not register.id is empty!')
    await this.userRepository.create(id)
    const { data } = await this.userRepository.login(id)
    if (!data?.id) throw new Error('invalid user')
    return data
  }

  @Action({ commit: 'setUser', rawError: true })
  public async addLike(payload: { postId: string }) {
    const clone = cloneDeep(this.user!)
    clone.likes.push({
      userId: clone.id,
      likedAt: new Date().getTime(),
      postId: payload.postId,
    })
    const { data } = await this.userRepository.edit(clone)
    return data
  }

  @Action({ commit: 'setUser', rawError: true })
  public async removeLike(payload: { postId: string }) {
    const clone = cloneDeep(this.user!)
    clone.likes = clone.likes.filter(
      (like: Like) => payload.postId !== like.postId
    )
    const { data } = await this.userRepository.edit(clone)
    return data
  }
}
