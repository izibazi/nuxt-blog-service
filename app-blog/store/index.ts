import { Store } from 'vuex'
import { initializeStore } from '~/utils/store-accessor'
const initalizer = (store: Store<any>) => initializeStore(store)
export const plugins = [initalizer]
export * from '~/utils/store-accessor'
