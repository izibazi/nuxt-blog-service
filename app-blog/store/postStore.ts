import { VuexModule, Module, Action, Mutation } from 'vuex-module-decorators'
import { IPost } from '~/domain/post'
import { $axios } from '~/utils/api'

@Module({ stateFactory: true, name: 'postStore', namespaced: true })
export default class PostStore extends VuexModule {
  posts: IPost[] = []

  @Mutation
  setPosts(posts: IPost[]) {
    this.posts = posts
  }

  @Mutation
  addPost(post: IPost) {
    this.posts.push(post)
  }

  @Action({ rawError: true, commit: 'addPost' })
  public async post(payload: { title: string; body: string; userId: string }) {
    const user = await $axios.$get(`/users/${payload.userId}.json`)
    console.log(user.id)
    if (user == null || user.id == null) {
      throw new Error('not found user')
    }
    const result = await $axios.$post('/posts.json', payload)
    const createdAt = new Date().getTime()
    const post = { id: result.name, ...payload, createdAt }
    const putData = {
      id: result.name,
      body: payload.body,
      title: payload.title,
      createdAt,
    }
    await $axios.$put(`/users/${user.id}/posts.json`, [
      ...(user.posts || []),
      putData,
    ])
    return post
  }
}
