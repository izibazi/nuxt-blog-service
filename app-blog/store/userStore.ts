import { VuexModule, Module, Action, Mutation } from 'vuex-module-decorators'
import { User } from '~/domain/user'
import { $axios } from '~/utils/api'

@Module({ stateFactory: true, namespaced: true, name: 'userStore' })
export default class UserStore extends VuexModule {
  users: User[] = []

  get user() {
    return (id: string): User | undefined => {
      return this.users.find((user) => user.id === id)
    }
  }

  @Mutation
  addUser(user: User) {
    this.users.push(user)
  }

  @Mutation
  setUsers(users: User[]) {
    this.users = users
  }

  @Action({ rawError: true, commit: 'addUser' })
  public async fetchUser(id: string) {
    if (this.user(id)) return
    const data = await $axios.$get(`/users/${id}.json`)
    return data
  }

  @Action({ rawError: true, commit: 'setUsers' })
  public async fetchUsers() {
    const data = await $axios.$get(`/users.json`)
    const users = Object.entries(data).map(([_, user]) => {
      return user
    })
    return users
  }
}
