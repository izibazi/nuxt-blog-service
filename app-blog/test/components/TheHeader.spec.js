import { mount, RouterLinkStub } from '@vue/test-utils'
import TheHeader from '~/components/TheHeader.vue'
import { createStore } from '~~/.nuxt/store'
import { initializeStore } from '~/utils/store-accessor'

describe('TheHeader', () => {
  let wrapper
  beforeEach(() => {
    initializeStore(createStore())
    wrapper = null
    wrapper = mount(TheHeader, {
      stubs: {
        NuxtLink: RouterLinkStub,
      },
    })
  })

  test('title', () => {
    expect(wrapper.findAll('.el-menu-item').at(0).props().route.path).toBe('/')
    expect(wrapper.find('h1').text()).toBe('Nuxt Diary App')
  })
  test('スナップショット', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
