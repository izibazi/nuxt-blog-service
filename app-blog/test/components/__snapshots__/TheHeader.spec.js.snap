// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheHeader スナップショット 1`] = `
<div>
  <transition-stub
    mode="out-in"
  >
    <ul
      class="el-menu--horizontal el-menu"
      role="menubar"
      tabindex="0"
    >
      <li
        class="el-menu-item"
        role="menuitem"
        style="border-bottom-color: transparent;"
        tabindex="-1"
      >
        <h1>
          Nuxt Diary App
        </h1>
      </li>
      <li
        class="el-menu-item"
        role="menuitem"
        style="border-bottom-color: transparent;"
        tabindex="-1"
      >
        投稿
      </li>
      <li
        class="el-menu-item"
        role="menuitem"
        style="border-bottom-color: transparent; float: right;"
        tabindex="-1"
      >
        ログイン
      </li>
      <li
        class="el-menu-item"
        role="menuitem"
        style="border-bottom-color: transparent; float: right;"
        tabindex="-1"
      >
        新規投稿
      </li>
    </ul>
  </transition-stub>
</div>
`;
